src/js/engagement_view/src/components/engagementView/sidebar/utils/toggleLensTable.tsx

```diff
@@ -0,0 +1,112 @@
+import React, { useEffect, useState } from "react";
+
+import Button from "@material-ui/core/Button";
+import KeyboardArrowDownOutlinedIcon from '@material-ui/icons/KeyboardArrowDownOutlined';
+import KeyboardArrowUpOutlinedIcon from '@material-ui/icons/KeyboardArrowUpOutlined';
+
+import Divider from "@material-ui/core/Divider";
+import Backdrop from "@material-ui/core/Backdrop";
+import CircularProgress from "@material-ui/core/CircularProgress";
+
+import { lensTable } from "./lensTable";
+import { getLenses } from "services/graphQLRequests/getLenses";
+
+import { ToggleLensTableProps, ToggleLensTableState } from "types/CustomTypes";
+
+import { useStyles } from "../styles";
+
+const defaultToggleLensTableState = (): ToggleLensTableState => {
+	return {
+		toggled: true,
+		lenses: [],
+		first: 100, // first is the page size
+		offset: 0, // by default, start from page 0
+	};
+};
+
+export function ToggleLensTable({ setLens }: ToggleLensTableProps) {
+	const classes = useStyles();
+
+	const [lensRetrievedState, setLensRetrievedState] = useState(null);
+	const [toggleTableState, setToggleTableState] = useState(
+		defaultToggleLensTableState()
+	);
+	const [pageState, setPageState] = useState(0);
+	const [rowsPerPageState, setRowsPerPageState] = useState(10);
+
+	const handleChangePage = (
+		event: React.MouseEvent<HTMLButtonElement, MouseEvent> | null,
+		page: number
+	) => {
+		setPageState(page);
+	};
+
+	const handleChangeRowsPerPage = (
+		event: React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement>
+	) => {
+		setRowsPerPageState(parseInt(event.target.value, 10));
+		setPageState(0);
+	};
+
+	useEffect(() => {
+		const interval = setInterval(() => {
+			getLenses(toggleTableState.first, toggleTableState.offset).then(
+				(response) => {
+					if (response.lenses && response.lenses !== toggleTableState.lenses) {
+						const lenses = toggleTableState.lenses.concat(response.lenses);
+						setLensRetrievedState(lenses as any);
+						setToggleTableState({
+							...toggleTableState,
+							offset: toggleTableState.offset + response.lenses.length || 0,
+							lenses,
+						});
+					}
+				}
+			);
```

Comment:
Let's create a `getAllLenses` function that pages in all of the lenses in one function. Then, once that function completes, let's set the state's offset to 0.

Currently I believe that if a lens is added to our graph there's the possibility that our paging logic will miss it.

Reply:
Fixing with a new issue. See: https://github.com/grapl-security/issue-tracker/issues/279